부산대 BE_오선재 Step3 - 배포하기

5seonjae opened the pull request:
### 배포된 URL : http://13.209.131.194:8080

## 변경 내용 요약

- WebConfig 에 CORS 매핑 추가

  - 모든 엔드포인트(/**)에 대해

    - 로컬 개발 서버 ( `http://localhost:3000` ) 에서의 AJAX 요청 허용

    - GET/POST/PUT/DELETE/OPTIONS/HEAD 메서드 허용

    - 자격 증명 쿠키 전송 ( `allowCredentials(true)` )

    - Pre-flight 캐시 ( 30분 ) 설정

- GitHub Actions 워크플로우 개선

  - build 잡

    - secrets 에서 JWT·Kakao 관련 환경변수 주입

    - Gradle 캐시, JAR 빌드 후 아티팩트 업로드

  - deploy 잡

    - SSH 로 EC2 에 scripts/, build/, app/ 디렉터리 생성

    - 빌드 아티팩트 다운로드 → EC2 에 `deploy.sh` 스크립트·JAR 파일 복사

    - `appleboy/ssh-action` 의 `envs:` 옵션으로 필요한 환경변수만 원격 셸에 `export`

    - `deploy.sh` 실행으로 애플리케이션 재시작

- 배포 스크립트(scripts/deploy.sh) 개선

  - build/libs 에서 -plain.jar 제외한 실행 가능한 fat-jar 자동 선택

  - 기존 프로세스 PID 감지→ SIGTERM → 5초 대기 후 종료

  - 새 JAR 복사·cd ~/app → nohup java -jar (--spring.profiles.active=prod)

  - 커맨드라인 인자로 `--jwt.secret`, `--kakao.client-id` 등 주요 시크릿/엔드포인트 주입 → Spring 플레이스홀더 해석 보장

  - 표준 출력·오류를 `app/app.log` 로 리다이렉트

---

## 기능 동작 확인

https://www.notion.so/Step2-23e75546dc3080e2a23af3ca2d663eef?source=copy_link

이전 2단계 미션에서 공유드렸던 페이지 링크에 배포 후 기능 동작 확인 시연 영상을 올렸습니다.

---

## 이번 단계를 하면서 느낀 점

이번에 처음으로 배포를 시도하면서, 배포 방법이 매우 다양하다는 것을 깨달았습니다.
특히 GitHub Actions 를 이용한 EC2 자동 배포 과정을 접하고, 셸 스크립트 문법부터 환경 변수 전달 방식까지 하나하나 익히느라 정말 많은 어려움을 겪었던 것 같습니다.
환경 변수를 원격 셸에 어떻게 전달해야 하는지 몰라서 불필요한 커밋을 잔뜩 남기기도 했고, 배포 상태를 확인하려면 푸시가 필요하다는 점도 처음 알게 되었습니다.
이 과정에서 코드 리뷰 일정에 차질이 생기실 수도 있을텐데, 양해 부탁드립니다.
아직 배포에 대한 지식이 부족하다고 느끼지만, **“일단 성공시키겠다”**는 목표로 달려들며 많이 성장한 것 같습니다.
앞으로도 부족한 부분을 꾸준히 학습하고, 배포 자동화와 운영 경험을 계속 쌓아나가겠습니다.

감사합니다!

5seonjae committed:
feat(product): Product 조회 시 옵션 연관 엔티티 즉시 패치 적용

- `ProductRepository.findById(Long)` 메서드에 `@EntityGraph(attributePaths = "options")` 애노테이션 추가
- Lazy 로딩으로 발생하던 N+1 문제 예방 및 조회 성능 최적화

5seonjae committed:
feat(auth): MemberExtractor 서비스 도입 및 로그인 사용자 추출 로직 통합

- `MemberExtractor` 컴포넌트 추가
    - HTTP 요청 헤더에서 Bearer 토큰 파싱 (`tokenExtractor.extractBearerHeader` → `bearerAuthHeaderParser.extractBearerToken`)
    - JWT Claims 검증 및 `memberRepository` 조회로 회원 객체 반환
- `LoginMemberArgumentResolver.resolveArgument`에서 직접 토큰 파싱 로직 제거 후 `memberExtractor.extractMember(req)` 호출로 단일 책임 원칙 준수
- `KakaoTokenInterceptor.preHandle`에서도 `memberExtractor.extractMember(req)` 사용해 중복 로직 통합
- `LoginMemberArgumentResolver.resolve` 제거
-  `WishControllerTest` 테스트 `setUp()` 스텁 로직 업데이트:
    - `loginMemberArgumentResolver.resolve` 스텁 제거 및 `kakaoTokenInterceptor.preHandle` 스텁 추가

5seonjae committed:
fix(auth): KakaoTokens.from 메서드에서 LocalDateTime.now() 호출 중복 제거

- `KakaoTokens.from(KakaoTokenResponseDto)` 내부에서 `LocalDateTime.now()`를 한 번만 호출하도록 `now` 변수로 추출
- `now.plusSeconds(...)` 및 `now.plusDays(...)` 방식으로 토큰 만료일 계산 일관성 강화

5seonjae committed:
refactor(entity): Order/Option 엔티티 유효성 검증 로직 일원화

- **Order 엔티티**
    - `quantity`, `message` 필드에 `@Min`, `@Max`, `@Size` 애노테이션 적용하여 Bean Validation 활용
    - 기존 `validate(...)` 메서드 및 생성자 내 호출 제거

- **Option 엔티티**
    - 필드 기반 검증 애노테이션(`@NotBlank`, `@Size`, `@Pattern`, `@Min`, `@Max`) 제거
    - 생성자 내부 `validate(product, name, quantity)` 메서드로 수동 검증 로직 통합

5seonjae committed:
refactor(auth): KakaoOAuthService·KakaoMessageService DI 방식을 생성자 주입으로 통일

- `KakaoOAuthService` 필드 `@Value` 애노테이션 제거 후 생성자 파라미터로 이동하여 `final` 필드 초기화
- `KakaoMessageService`에서 `apiUrl`, `templateId` 필드의 `@Value` 애노테이션을 생성자 파라미터로 옮겨 DI 방식 일원화
- `WebClient`, `ObjectMapper` 의존성도 생성자 주입으로 변경하여 가독성 및 테스트 편의성 향상

5seonjae committed:
refactor(message): sendOrderMemo 결과 코드 처리 상수화 및 타입 안정성 강화

- 응답 성공 코드 `0`을 `KAKAO_SUCCESS_CODE` 상수로 추출하여 매직 넘버 제거
- `resp.get("result_code")`를 `Integer resultCode`로 캐스팅해 타입 안정성 보강
- 상수와 비교하도록 변경하여 응답 검증 로직 가독성 및 유지보수성 향상

5seonjae committed:
fix(test): 주문 삭제 테스트 로직 수정

- `OrderService.deleteOrderForMember`가 `orderRepository.delete(order)`를 호출하도록 변경됨에 따라
- 테스트의 `verify(orderRepository).delete(orderId)` → `verify(orderRepository).delete(order)`로 검증 대상 수정
- `Optional.of(new Order(...))` 객체를 로컬 변수 `order`에 할당한 후 해당 변수를 이용해 `verify` 호출하도록 코드 보강

5seonjae committed:
fix(test): @CsvSource 주석 제거 및 테스트 코드 간결화

- `getOrderList_success_multiple` 메서드의 `// page, quantity, message, isFirst, hasPrevious, hasNext` 주석 삭제
- `@CsvSource`만으로도 테스트 시나리오가 명확히 파악되도록 코드 간결화 및 가독성 향상

5seonjae committed:
feat(readme): Step3 ‘배포하기’ 가이드 및 CI/CD 파이프라인 문서화

- GitHub Actions 워크플로우(deploy.yml) 작성 방법 추가
    - main/develop 브랜치 푸시 시 빌드 → EC2 배포 과정 설명
    - appleboy/scp-action, ssh-action 활용 예시 포함
- 서버 재시작 스크립트(scripts/deploy.sh) 예시 추가
    - 기존 JAR 프로세스 종료 → 신규 JAR 복사 및 백그라운드 실행 순서 명시
- 글로벌 CORS 설정 섹션 보완
    - `/api/**` 엔드포인트 대상, 허용 메서드·도메인·인증 설정 예시
- HTTPS 적용 안내 (Let’s Encrypt + Nginx) 추가
    - Nginx 설치, Certbot 설정 명령어 예시 삽입
- 시스템 개요 및 요구 사항 정리
    - EC2 환경, 도메인 A-레코드, GitHub Secrets, Gradle Wrapper 등 사전 준비 항목 정리
    - 전체 아키텍처 다이어그램 부연 설명

위 커밋으로 배포 단계에 필요한 전반적인 CI/CD 및 보안 설정 가이드를 README.md에 상세히 문서화했습니다.

5seonjae committed:
chore(config): 프로덕션 프로파일용 application-prod.properties 추가

- H2 콘솔 비활성화 설정 추가 (spring.h2.console.enabled=false)
- Hibernate SQL 포맷팅 및 출력 비활성화 설정 추가
    - spring.jpa.properties.hibernate.format_sql=false
    - spring.jpa.show-sql=false
- DDL-auto 모드를 validate로 설정 (spring.jpa.hibernate.ddl-auto=validate)

5seonjae committed:
feat(deploy): 프로덕션 배포 스크립트(scripts/deploy.sh) 추가

- 빌드된 JAR 자동 복사 및 배포 로직 구현
- 기존 실행 중인 애플리케이션 프로세스(SIGTERM) 종료 후 대기
- 새 JAR 파일 복사 → `nohup java -jar` 백그라운드 실행
- `--spring.profiles.active=prod` 옵션으로 prod 프로파일 활성화

5seonjae committed:
feat(ci): GitHub Actions 기반 자동 배포 파이프라인 추가

- step3 브랜치 push 트리거 시 Gradle 빌드 후 JAR 아티팩트 업로드
- appleboy/scp-action으로 EC2 ~/build/ 경로에 JAR 전송
- appleboy/ssh-action으로 원격 deploy.sh 실행 → 서비스 재시작
- concurrency(group) 설정으로 동일 브랜치 중복 배포 방지

5seonjae committed:
chore(ci): deploy.yml에 GitHub Actions 환경변수 추가

- JWT_SECRET 및 JWT_EXPIRATION_MS Secrets 매핑
- KAKAO_CLIENT_ID, KAKAO_AUTH_URL, KAKAO_API_URL, KAKAO_REDIRECT_URI, KAKAO_TEMPLATE_ID Secrets 매핑

5seonjae committed:
chore(config): application.properties에 optional .env 프로퍼티 import 설정 추가

- `spring.config.import=optional:file:.env[.properties]` 설정 추가

5seonjae committed:
chore(ci): deploy.yml에 스크립트 전송 단계 추가 및 전송 로직 개선

- `deploy` job에 리포지토리 체크아웃(`actions/checkout`) 단계 명시
- artifacts 다운로드 단계를 `Download artifact`로 분리
- `scripts/deploy.sh` 파일을 EC2의 `/home/<user>/scripts/`로 복사하는 단계 추가
- 빌드된 JAR 전송 단계를 `/home/<user>/build/`로 분리
- SSH 실행 스크립트 호출 시 경로를 `~/scripts/deploy.sh`로 반영

5seonjae committed:
chore(ci): deploy.yml에 원격 디렉토리 준비 단계 추가

- `appleboy/ssh-action`를 사용해 `/home/${{ secrets.EC2_USER }}/scripts` 및 `/home/${{ secrets.EC2_USER }}/build` 디렉터리 사전 생성
- 스크립트 및 JAR 복사 전 환경을 준비해 배포 안정성 강화

5seonjae committed:
chore(ci): deploy.yml Download artifact 경로 명시

- `actions/download-artifact` 단계에 `path: build/` 옵션 추가
- 아티팩트 다운로드 위치를 `build/` 디렉터리로 고정해 후속 복사 로직 안정화

5seonjae committed:
Update deploy.yml

5seonjae committed:
Update deploy.yml

5seonjae committed:
Update deploy.yml

5seonjae committed:
Update deploy.yml

5seonjae committed:
Update deploy.sh

5seonjae committed:
fix deploy.sh

5seonjae committed:
Fix deploy.sh

5seonjae committed:
fix(application-prod.properties) - `spring.jpa.hibernate.ddl-auto=validate` 제거

5seonjae committed:
fix(deploy.sh) - 환경 변수 추가

5seonjae committed:
fix(deploy.yml) - 환경 변수 설정

5seonjae committed:
fix(deploy.yml) - 환경 변수 관련 수정

5seonjae committed:
fix(deploy.yml) - 환경 변수 관련 수정

5seonjae committed:
fix(deploy.yml) - 환경 변수 설정

5seonjae committed:
fix(deploy.yml) - 환경 변수 설정

5seonjae committed:
fix(deploy.sh) - 환경 변수 설정

5seonjae committed:
fix(deploy) - 환경 변수 설정 수정

5seonjae committed:
fix(deploy) - 환경 변수 설정 수정

5seonjae committed:
fix(kakao): ResponseCookie에서 secure 옵션 제거

- `ResponseCookie.from("AUTH", jwt)` 생성 시 `.secure(true)` 호출 부분 삭제
- HTTP 환경(로컬 개발)에서도 쿠키가 정상 전송되도록 수정

5seonjae committed:
feat(cors): 전역 CORS 매핑 설정 추가

- `CorsRegistry`에 `/**` 패턴 허용
- `http://localhost:3000` Origin 허용
- GET, POST, PUT, DELETE, OPTIONS, HEAD 메서드 허용
- 모든 헤더 허용 및 자격 증명 허용(`allowCredentials(true)`)
- preflight 캐시 지속시간 1800초 설정(`maxAge(1800)`)

5seonjae committed:
docs(README.md): 체크리스트 반영

5seonjae committed:
fix(auth): 토큰 만료 버퍼 상수화 및 isAccessExpired 로직 개선

- `TOKEN_REFRESH_BUFFER(Duration.ofSeconds(60))` 상수 추가로 버퍼 시간 매직 넘버 제거
- `isAceessExpired()` 메서드에서 하드코딩된 `plusSeconds(60)` → `plus(TOKEN_REFRESH_BUFFER)` 사용으로 가독성·유지보수성 강화

Gyuchool commented on README.md:
```diff
@@ -210,4 +210,176 @@ sequenceDiagram
+                                     ssh + scp │ 
+┌────────────────────────┐           ┌─────────▼───────────┐
+│       EC2 (Ubuntu)     │           │    deploy.sh runs   │
+│ Nginx (TLS termination)│◀──https──▶│   Spring Boot JAR   │
```
그림이 약간 이상한거 같습니다! jar가 ec2에 올라가서 실행되는게 아닐까요?
https로 통신한다고 작성해주신 이유가 궁금합니다 :)

5seonjae replied:
리뷰 감사합니다!
지적해주신 대로 다이어그램이 실제 배포 흐름과 맞지 않네요. ( 까먹고 수정을 안했던 것 같습니다;; ㅎㅎ )

현재 구현된 구조는 아래와 같습니다.

- GitHub Actions

  - push → 빌드 → ssh + scp로 EC2에 JAR 파일 전송

- EC2 (Ubuntu)

  - `/scripts/deploy.sh` 실행 → Spring Boot JAR 실행

초기 설계 단계에서 TLS termination ( Nginx → HTTPS ) 구성을 고려하며 다이어그램에 포함시켰지만, 실제로는 아직 적용하지 않은 상태입니다.
README의 해당 부분을 삭제하고, 위 순서대로 배포 흐름을 표현하도록 바로 수정하겠습니다!
감사합니다. 😊

Gyuchool commented on README.md:
```diff
@@ -210,4 +210,176 @@ sequenceDiagram
+
+## 🚀 Step 3 배포하기
+
+지금까지 만든 선물하기 서비스를 **GitHub Actions → EC2** 파이프라인으로 자동 배포하고 클라이언트가 `https://gift.example.com`에서 안전하게 API를 호출하도록 구성합니다.
```
https://gift.example.com 도메인은 아직 연결 안된걸까요? 임시로 작성해둔것 맞을까요?

5seonjae replied:
지적 감사합니다!
`https://gift.example.com` 도메인은 배포 전에 `README.md` 를 작성하면서 임시로 만들어 둔 placeholder 라 실제 도메인과 다릅니다.
실제 도메인으로 수정하도록 하겠습니다! 😊

Gyuchool commented on README.md:
```diff
@@ -210,4 +210,176 @@ sequenceDiagram
 
+---
+
+## 🚀 Step 3 배포하기
```
정리를 잘해주셔서 무슨 작업을 하셨는지 이해 쏙쏙이네요!👍

Gyuchool commented on scripts/deploy.sh:
```diff
@@ -0,0 +1,32 @@
+
+if [ -n "$PID" ]; then
+echo "▶ stop running app (pid=$PID)"
+kill -15 "$PID"
```
kill -15로 설정하신 이유가 있으실까요? -9와는 어떤 차이가 있을까요?

5seonjae replied:
리뷰 감사합니다!

`kill -15` ( SIGTERM ) 는 프로세스에 “정상 종료” 요청을 보내는 시그널이라, Spring Boot 가 제공하는 shutdown hook 을 실행해 현재 수행 중인 작업을 정리 ( 종료 전 리소스 해제, 로그 flush 등 ) 할 수 있습니다.

반면 `kill -9` ( SIGKILL ) 는 커널이 바로 프로세스를 강제 종료시켜 cleanup 로직이 전혀 실행되지 않으므로, 파일 잠금 해제나 트랜잭션 롤백 등이 누락될 위험이 있습니다.

따라서 안전한 종료와 서비스 무결성 유지를 위해 우선 SIGTERM 을 보내고, 만약 프로세스가 일정 시간 안에 종료되지 않을 경우 추가로 SIGKILL 을 보내는 패턴으로 사용할 수 있습니다. 감사합니다! 😊

Gyuchool replied:
graceful shutdown 이란 개념과 관련지으면 좋습니다!

Gyuchool commented on src/main/java/gift/config/WebConfig.java:
```diff
@@ -39,4 +40,14 @@ public void addInterceptors(InterceptorRegistry registry) {
+            .allowedMethods("GET","POST","PUT","DELETE","OPTIONS","HEAD")
+            .allowedHeaders("*")
+            .allowCredentials(true)
+            .maxAge(1800);
```
maxAge를 1800초로 설정하신 이유가 있을까요?🙄

5seonjae replied:
리뷰 감사합니다!

`maxAge` 는 브라우저가 Pre-flight 요청 ( OPTIONS ) 의 결과를 캐싱해 두는 기간 ( 초 ) 을 의미합니다.

`maxAge` 를 **1800초 ( 30분 )** 으로 설정한 이유는 다음과 같습니다.

- 성능 최적화

  - 같은 출처 ( `http://localhost:3000` ) 에서 반복해서 API 를 호출할 때 30분 동안은 추가적인 Pre-flight 요청 없이 실제 요청만 실행되어, 개발 환경에서 서버 부하를 줄일 수 있습니다.

- 유연성 고려

  - 30분 정도면 CORS 정책이 변경될 때도 비교적 빠르게 적용을 확인할 수 있는 적당한 타임윈도우입니다.

필요에 따라 더 짧게 ( 예 : 600초 ) 설정하면 정책 변경을 즉시 반영할 수 있고, 더 길게 ( 예 : 3600초 이상 ) 설정하면 Pre-flight 호출 빈도를 더욱 줄여 성능 이점을 키울 수 있습니다.

현재 개발 단계에서는 1800초가 무난한 절충점이라 선택했고, 운영 환경이나 정책 변경 빈도에 따라 얼마든지 조정 가능합니다.

Gyuchool replied:
적절히 고민하시고 설정하셨군요👍 한발 더 나아가 추가 질문 드려봅니다!

CORS 설정에 maxAge를 설정하면 브라우저가 Preflight 응답을 일정 시간 캐싱하게 되는데요,
만약 배포 시에 특정 Origin을 CORS 허용 목록에 추가하는 것을 깜빡했다면, 초기에는 캐시 덕분에 문제가 드러나지 않아 정상 동작하는 것처럼 보일 수 있을 것 같습니다.

하지만 시간이 지나 maxAge가 만료되면 브라우저가 다시 Preflight 요청을 보내게 되고, 이때야 비로소 CORS 에러가 발생하면서 문제가 수면 위로 드러날 수 있을 텐데요.

이런 상황은 어떻게 하면 미리 방지하거나 안전하게 대응할 수 있을까요? 🙄

5seonjae replied:
날카로운 지적 감사합니다! 🙏

지적하신 대로, `maxAge` 덕분에 초기에 CORS 에러가 보이지 않다가 만료 시점에 갑자기 수면 위로 드러나는 상황이 발생할 수 있습니다. 
이런 경우를 예방하거나 안전하게 대응하는 방안으로는 크게 세 가지를 고려해 볼 수 있습니다.

### 1. 캐시 기간을 짧게 두거나 무효화 전략 사용

- 짧은 `maxAge` ( 예 : 300초 = 5분 ) 로 설정해, 변경된 정책이 빠르게 반영되도록 합니다.
- Pre-flight 응답에 `Cache-Control: no-cache, must-revalidate` 헤더를 추가해 브라우저가 항상 서버와 재검증하도록 유도할 수도 있습니다.

```java
@Bean
public WebFilter corsHeaderFilter() {
  return (req, res, chain) -> {
    ServerHttpResponse response = (ServerHttpResponse) res;
    response.getHeaders().setCacheControl("no-cache, must-revalidate");
    return chain.filter(req, res);
  };
}
```

→ 이러면 `maxAge` 만료 후에도 즉시 최신 정책을 적용하게끔 클라이언트 캐시를 무효화할 수 있습니다.

### 2. 허용 Origin 목록 관리 프로세스 자동화

- 배포 파이프라인에 “새로운 Origin 추가 → CORS 설정 갱신 → 배포” 단계를 포함시키고, 변경 직후 서버가 재시작되면 새로운 설정이 바로 적용되도록 합니다.
- 운영 중에도 외부 설정 파일 ( YAML/DB 등 ) 에 Origin 목록을 갖다 쓰는 구조로 짜두면, 서버 재시작 없이도 동적으로 새로운 도메인을 허용할 수 있습니다.

### 3. 클라이언트 안내 & 자체 테스트

- 정책 변경 ( 새 Origin 허용 ) 시, 배포 노트나 슬랙 등으로 “CORS 캐시가 최대 maxAge 만큼 남아있을 수 있으니, 테스트 전 브라우저 캐시 ( 개발자 도구 → Disable cache 등 ) 를 비활성화 해주세요” 를 안내합니다.
- QA 환경에서는 `maxAge` 를 0으로 두고 매번 preflight 를 날려보는 테스트를 자동화해 놓으면, 운영 전 에러 가능성을 미리 잡아낼 수 있습니다.

정책 변경이 잦지 않다면 짧은 `maxAge` + 캐시 무효화 헤더 조합을 적용하고, 자주 도메인 추가가 필요하다면 외부 설정 ( 파일 / DB ) 기반의 동적 CORS 로직을 적용하면 문제가 터지기 전에 미리 방지하면서도 운영 편의성을 모두 챙길 수 있을 것 같습니다!

Gyuchool commented on src/main/java/gift/entity/Option.java:
```diff
@@ -54,6 +40,8 @@ public class Option {
     protected Option() {}
 
     public Option(Long id, Product product, String name, int quantity) {
+        validate(product, name, quantity);
```
단위 테스트를 수행할때 더욱 편해지겠군요👍

Gyuchool commented on src/main/java/gift/entity/Option.java:
```diff
@@ -64,6 +52,29 @@ public Option(Product product, String name, int quantity) {
+        if (name.length() > 50) {
+            throw new IllegalArgumentException("최대 50자까지 가능합니다.");
+        }
+        if (!name.matches("^[a-zA-Z0-9가-힣()\\[\\]+\\-&/_ ]*$")) {
```
패턴 매칭은 비싼 연산이여서 캐싱해두면 어떨까 싶네요!

5seonjae replied:
리뷰 감사합니다!
`String#matches` 는 내부적으로 매번 `Pattern.compile` 을 실행하기 때문에 반복 호출 시 성능에 비용이 발생할 수 있습니다. 
제안해주신 대로, 미리 컴파일한 Pattern을 재사용하도록 아래와 같이 수정해보겠습니다.

```java
@Entity
@Table(name = "options")
public class Option {
    // … 생략 …

    // ① static final로 미리 컴파일
    private static final Pattern NAME_PATTERN =
        Pattern.compile("^[a-zA-Z0-9가-힣()\\[\\]+\\-&/_ ]*$");

    private void validate(Product product, String name, int quantity) {
        // … 기존 검증 로직 …
        if (!NAME_PATTERN.matcher(name).matches()) {
            throw new IllegalArgumentException(
                "유효한 특수문자 ( '( )', '[ ]', '+', '-', '&', '/', '_' ) 만 사용할 수 있습니다."
            );
        }
        // … 나머지 검증 …
    }

    // … 이하 생략 …
}
```

이렇게 변경하면 패턴 매칭 비용을 크게 줄일 수 있어 성능 최적화에 도움이 될 것 같습니다. 
다시 한 번 좋은 제안 감사드립니다! 😊

Gyuchool commented on src/main/java/gift/entity/Option.java:
```diff
@@ -64,6 +52,29 @@ public Option(Product product, String name, int quantity) {
+        if (name == null || name.trim().isEmpty()) {
+            throw new IllegalArgumentException("옵션 이름은 필수입니다.");
+        }
+        if (name.length() > 50) {
```
50의 의미를 알기 어렵습니다! 매직넘버는 모두 제거해주세요😊

5seonjae replied:
리뷰 감사합니다!
이 부분은 이전에 매직 넘버를 상수로 분리하면서 놓쳤네요.
Option 클래스에 `NAME_MAX_LENGTH = 50` 상수를 추가하고, `name.length() > NAME_MAX_LENGTH` 로 변경해서 반영하겠습니다.
다른 코드들도 살펴보고 매직 넘버 관련 부분은 최대한 수정해보도록 하겠습니다.
빠르게 수정하겠습니다! 😊

Gyuchool commented on src/main/java/gift/service/KakaoMessageService.java:
```diff
@@ -22,25 +22,22 @@
     private final WebClient kakaoClient;
     private final ObjectMapper objectMapper;
+    private final Long templateId;
+    private static final int KAKAO_SUCCESS_CODE = 0;
```
자바 관례로 상수를 맨 위에 적곤합니다:)

Gyuchool commented on src/main/java/gift/service/KakaoMessageService.java:
```diff
@@ -56,7 +53,8 @@ public void sendOrderMemo(Order order, String accessToken) {
                 .block();
 
-            if (!Objects.equals(resp.get("result_code"), 0)) {
+            Integer resultCode = (Integer) resp.get("result_code");
```
응답에 result_code값이 없으면 어떻게 되나요?

5seonjae replied:
리뷰 감사합니다!

응답에 `result_code` 값이 없으면 `Integer resultCode = (Integer) resp.get("result_code");` 에서 `resp.get("result_code")` 가 null 을 반환해 `resultCode` 는 null 이 됩니다.

이후에 
```java
if (!Objects.equals(resultCode, KAKAO_SUCCESS_CODE)) {
    throw new ExternalServiceException("카카오톡 전송 실패: " + resp);
}
```
에서 `Objects.equals(null, 0)` 는 false 로 평가되므로 if 문이 실행되어 `ExternalServiceException` 이 발생합니다.
즉, `result_code` 가 누락된 경우에도 예외를 던져 “전송 실패”로 처리되도록 의도된 흐름입니다.

필요하다면, 누락 시 기본값을 지정하거나 ( `Map#getOrDefault` ), resp 자체가 null 인지 먼저 검증하는 로직을 추가해도 좋을 것 같습니다.
감사합니다! 😊

Gyuchool commented on README.md:
```diff
@@ -210,4 +210,176 @@ sequenceDiagram
+| 항목 | 설명                                              |
+|------|-------------------------------------------------|
+|EC2| Ubuntu 22.04, OpenJDK 21, 포트 `8080` 오픈          |
+|도메인| `gift.example.com` A-레코드 → EC2 IP               |
```
A레코드와 CNAME의 차이는 무엇일까요?

5seonjae replied:
지적 감사합니다!
현재는 도메인 연결을 아직 진행하지 않아 해당 A레코드/CNAME 설명을 README 에서 과감히 삭제했습니다.
실제 도메인 설정 시에는

- A레코드 : 도메인 → 서버 IP 직접 매핑

- CNAME : 도메인 → 다른 도메인 ( alias ) 매핑

로 사용하게 될 예정입니다.
필요해지면 위 내용을 참고해 다시 추가하겠습니다!

Gyuchool commented on src/main/java/gift/config/WebConfig.java:
```diff
@@ -39,4 +40,14 @@ public void addInterceptors(InterceptorRegistry registry) {
+    public void addCorsMappings(CorsRegistry registry) {
+        registry.addMapping("/**")
+            .allowedOrigins("http://localhost:3000")
+            .allowedMethods("GET","POST","PUT","DELETE","OPTIONS","HEAD")
```
OPTIONS, HEAD 메서드는 각각 언제 사용되나요?

5seonjae replied:
질문 감사합니다!

`OPTIONS` 와 `HEAD` 는 모두 HTTP 표준 메서드지만, 용도와 사용 시점이 다릅니다.

### OPTIONS

- CORS Pre-flight 요청

  - 브라우저가 “비단순 요청” ( 예 : Content-Type: application/json, 사용자 정의 헤더, PUT/DELETE 등 ) 을 보내기 전, 실제 요청을 보내도 안전한지 확인하기 위해 자동으로 {Origin, Access-Control-Request-Method, Access-Control-Request-Headers} 헤더를 담아 서버에 보냅니다.

- 서버 지원 메서드 조회

  - API 의 루트나 특정 리소스 경로에 `OPTIONS` 요청을 보내면, 서버가 Allow 헤더에 지원 가능한 메서드 목록 ( GET, POST, PUT …) 을 응답해 줍니다.

### HEAD

- 메타정보만 조회

  - GET 요청과 동일한 처리 로직을 거치지만, 응답 본문 ( body ) 은 전송하지 않고 헤더만 반환합니다.

- 리소스 유효성/변경 확인

  - 캐시 유효성 검사 ( Last-Modified, ETag 확인 )

  - 파일 사이즈 ( Content-Length ) 나 콘텐츠 타입 확인

    - 다운로드 전 “존재 여부”만 체크

이 두 메서드를 `allowedMethods("…","OPTIONS","HEAD")` 에 포함시키면, 브라우저의 CORS pre-flight( `OPTIONS` ) 와 클라이언트가 헤더만 확인할 때( `HEAD` ) 모두 허용하도록 서버가 응답하도록 설정됩니다.

Gyuchool commented on src/main/java/gift/config/WebConfig.java:
```diff
@@ -39,4 +40,14 @@ public void addInterceptors(InterceptorRegistry registry) {
+        registry.addMapping("/**")
+            .allowedOrigins("http://localhost:3000")
+            .allowedMethods("GET","POST","PUT","DELETE","OPTIONS","HEAD")
+            .allowedHeaders("*")
```
allowHeaders도 특정해주면 더 안정적인 서비스가 될것같네요👍

5seonjae replied:
좋은 제안 감사합니다!
현재는 모든 헤더를 허용하기 위해 `allowedHeaders("*")` 로 설정해두었는데, 실제로 사용하는 헤더만 명시하면 CORS 보안을 더 강화할 수 있습니다.

예를 들어, API 호출 시 주로 사용하는 헤더만 나열하도록 아래처럼 수정해보겠습니다.

```java
@Override
public void addCorsMappings(CorsRegistry registry) {
    registry.addMapping("/**")
        .allowedOrigins("http://localhost:3000")
        .allowedMethods("GET","POST","PUT","DELETE","OPTIONS","HEAD")
        .allowedHeaders(
            "Authorization",    // JWT 토큰
            "Content-Type",     // 요청 바디 타입
            "Accept",           // 응답 허용 타입
            "X-Requested-With"  // Ajax 요청 식별용
            // 필요 시 "Cookie" 등을 추가 가능
        )
        .allowCredentials(true)
        .maxAge(1800);
}
```

이렇게 하면 불필요한 헤더 요청을 차단해 CORS 정책을 더욱 엄격하게 적용할 수 있습니다. 
다시 한 번 유용한 피드백 감사합니다! 😊

Gyuchool reviewed (changes requested):
안녕하세요, 선재님!
마지막 미션도 잘 구현해주셨네요 💯

요구사항을 충실히 반영해주셔서 머지해도 무방하지만, 마지막 미션인 만큼 한 단계 더 고민하고 학습해보면 좋을 포인트가 있어 RC 드립니다 🙏
추가로 코멘트를 남겨두었으니 확인 후 반영해주시고, 의견 있으시면 함께 남겨주세요.

궁금한 점은 디스코드 DM이나 PR 코멘트로 언제든 편하게 문의주세요!

5seonjae committed:
docs: 배포 파이프라인 및 README.md 최신화

- 기존 HTTPS 및 Nginx 설정 내용 제거
- GitHub Actions → EC2 자동 배포 워크플로우(step3 브랜치) 설명 보강
  - `.github/workflows/deploy.yml` 구성:
    - Gradle 빌드 → artifact 업로드 → EC2 디렉토리 생성
    - `appleboy/scp-action` 으로 JAR 및 `deploy.sh` 전송
    - `appleboy/ssh-action` 으로 원격 서비스 재시작
- `scripts/deploy.sh` 배포 스크립트 예시 추가
    - 기존 실행 중인 프로세스 종료
    - 최신 JAR 복사 및 `nohup java -jar` 방식으로 프로덕션 프로필 실행
- 글로벌 CORS 설정 예시(`CorsConfig`) 포함
    - `/**` 경로 및 `http://localhost:3000` 도메인 허용

5seonjae committed:
refactor(option): 이름 검증 정규식 패턴 상수화

- validate 메서드 내 정규식 컴파일을 static final Pattern NAME_PATTERN으로 분리
- 반복 컴파일 제거로 성능 최적화 및 가독성 개선

5seonjae committed:
refactor(member): 이메일·비밀번호 검증 로직 상수화

- 이메일 형식 검사용 정규식 패턴을 `EMAIL_PATTERN` 상수로 분리
- 비밀번호 최소 길이를 `PASSWORD_MIN_LENGTH` 상수로 정의
- `validate` 메서드에서 하드코딩된 값 대신 상수 사용으로 가독성 및 유지보수성 향상

5seonjae committed:
refactor(cors): CORS allowedHeaders를 명시적 헤더 목록으로 제한

- allowedHeaders("*")를 아래 헤더 목록으로 변경
    - Authorization
    - Content-Type
    - Accept
    - X-Requested-With
    - Cookie
- CORS 보안 강화 및 불필요한 헤더 허용 방지

5seonjae committed:
docs: README.md에 글로벌 CORS 설정 예시 추가

- WebConfig 클래스 코드 반영
- addCorsMappings()를 통한 `/api/**` CORS 매핑 예시 포함
- 허용 도메인, 메서드, 헤더, 자격 증명 설정 안내

Gyuchool reviewed (approved):
안녕하세요 선재님!
이번에도 학습도 잘해주셨고 피드백 반영도 잘해주셨네요💯

마지막으로 고민해볼만한 포인트 남겨두었으니 학습해보시면 좋겠네요
마지막 미션은 여기서 병합하도록 하겠습니다:)
